excise/utils.py

```diff
@@ -0,0 +1,674 @@
+        TitleTransferCode="DEST",
+        TransactionType=transaction_type,
+        TransactionLines=lines,
+        UserTranId=user_tran_id,
+        Discount=discount,
+    )
+
+    return data
+
+
+def get_checkout_lines_data(
+    checkout_info: "CheckoutInfo",
+    lines_info: Iterable["CheckoutLineInfo"],
+    config: AvataxConfiguration,
+    discounted: bool = False,
+    discounts=None,
+) -> List[TransactionLine]:
+    data: List[TransactionLine] = []
+    channel = checkout_info.channel
+    tax_included = Site.objects.get_current().settings.include_taxes_in_prices
+    shipping_address = checkout_info.shipping_address
+    if shipping_address is None:
+        raise TaxError("Shipping address required for ATE tax calculation")
+    for line_info in lines_info:
+        append_line_to_data(
+            amount=base_calculations.base_checkout_line_total(
+                line_info, channel, discounts
+            ).net.amount,
+            data=data,
+            line_id=line_info.line.id,
+            quantity=line_info.line.quantity,
+            shipping_address=shipping_address,
+            tax_included=tax_included,
+            variant=line_info.line.variant,
+            variant_channel_listing=line_info.channel_listing,
+            discounted=discounted,
+        )
+
+    append_shipping_to_data(
+        data,
+        config.shipping_product_code,
+        shipping_address,
+        checkout_info.shipping_method_channel_listings,
+    )
+    return data
+
+
+def generate_request_data_from_checkout(
+    checkout_info: "CheckoutInfo",
+    lines_info: Iterable["CheckoutLineInfo"],
+    config: AvataxConfiguration,
+    transaction_token=None,
+    transaction_type=TRANSACTION_TYPE,
+    discounts=None,
+):
+    discount_amount = Decimal(checkout_info.checkout.discount_amount)
+    invoice_number = None
+    user_tran_id = None
+    discounted = True if discount_amount > 0 else False
+
+    # Do not discount product price
+    lines = get_checkout_lines_data(
+        checkout_info,
+        lines_info,
+        config,
+        discounts=None,
+        discounted=discounted,
+    )
+
+    data = generate_request_data(
+        transaction_type,
+        lines=lines,
+        invoice_number=invoice_number,
+        user_tran_id=user_tran_id,
+        discount=discount_amount
+    )
+    return data
+
+
+def get_order_lines_data(
+    order: "Order", discounted: bool = False, discounts=None
+) -> List[TransactionLine]:
+
+    data: List[TransactionLine] = []
+    order_lines = order.lines.all()
+
+    tax_included = Site.objects.get_current().settings.include_taxes_in_prices
+    shipping_address = order.shipping_address
+    if shipping_address is None:
+        raise TaxError("Shipping address required for ATE tax calculation")
+
+    for line in order_lines:
+        variant = line.variant
+        if variant is None:
+            continue
+
+        variant_channel_listing = line.variant.channel_listings.get(
```

Comment:
We fetch each `variant_channel_listing` separately from the database for all lines.  We need to optimize it.